### ISSUE TITLE
Incorrect implicit any when contextual type is union type

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#10048]
Distilled from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/types-2.0/yeoman-generator

Code (actual behavior)
```ts
interface IPromptOptions {
	filter: (input: any) => void;
}

interface Generator {
	prompt1(opt: IPromptOptions | Array<IPromptOptions>): void;

	prompt2(opt: IPromptOptions): void;
}

declare const generator: Generator;
// Error: 'input' implicitly has type 'any'
generator.prompt1({ filter: (input) => { } });
// No error
generator.prompt2({ filter: (input) => { } });
```
Expected behavior: No error on either call (previous behavior)

Seems to have been introduced with the literal types PR.

**Comment by ahejlsberg** [comment]
This is an effect of not performing subtype reduction on the `filter` properties of `IPromptOptions` and `Array<IPromptOptions>`. It so happened to reduce the contextual type to `(input: any) => void` before, whereas we now end up with a contextual type that is a union of the two function types, and since the signatures differ we don't provide a contextual type for the parameter. Thus, the parameter gets an implicit any type. I think this behavior is fine.
